common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -395,19 +395,17 @@ void deleteExecutorDirs(AppShuffleInfo appShuffleInfo) {
   @Override
   public StreamCallbackWithID receiveBlockDataAsStream(PushBlockStream msg) {
     AppShuffleInfo appShuffleInfo = validateAndGetAppShuffleInfo(msg.appId);
-    if (appShuffleInfo.attemptId != msg.appAttemptId) {
-      // If this Block belongs to a former application attempt, it is considered late,
-      // as only the blocks from the current application attempt will be merged
-      // TODO: [SPARK-35548] Client should be updated to handle this error.
-      throw new IllegalArgumentException(
-        String.format("The attempt id %s in this PushBlockStream message does not match "
-          + "with the current attempt id %s stored in shuffle service for application %s",
-          msg.appAttemptId, appShuffleInfo.attemptId, msg.appId));
-    }
     // Use string concatenation here to avoid the overhead with String.format on every
     // pushed block.
     final String streamId = OneForOneBlockPusher.SHUFFLE_PUSH_BLOCK_PREFIX + "_"
       + msg.shuffleId + "_" + msg.shuffleMergeId + "_" + msg.mapIndex + "_" + msg.reduceId;
+    if (appShuffleInfo.attemptId != msg.appAttemptId) {
+      // If this Block belongs to a former application attempt, it is considered late,
+      // as only the blocks from the current application attempt will be merged
+      throw new BlockPushNonFatalFailure(new BlockPushReturnCode(ReturnCode.TOO_OLD_ATTEMPT_PUSH.id(),
+        streamId).toByteBuffer(),
+        BlockPushNonFatalFailure.getErrorMsg(msg.appAttemptId, ReturnCode.TOO_OLD_ATTEMPT_PUSH));
```

Comment:
Nit: we would get different error message for this exception on the server side vs. the client side. Server side would say "App attempt XXX is a too old app attempt" while client side would say "Block XXX is a too old app attempt".
It's better to keep both consistent by having the error message describing the situation from the perspective of the individual shuffle block, e.g. "Block XXX is from an old app attempt".

Reply:
@Victsm Very good suggestion, i will change in latest PR.